### PR TITLE
fix(wrangler): Fix `pages dev` watch mode [_worker.js]

### LINKED ISSUE #4824: 🚀 Feature Request: Pages functions with _worker.js does not hot reload for any changes other than the _worker.js file
### Describe the solution

It would be nice for wrangler to reload on any changes in the directory and subdirectories. Currently it will only reload if _worker.js is changed. It won't reload if any of its dependencies are modified including files imported from the same directory.

Using:

```
npx wrangler pages dev ./public
```

### PR DESCRIPTION
**Pull request by CarmenPopoviciu**
## What this PR solves / how to test

The watch mode in `pages dev` for Advanced Mode projects is currently partially broken, as it only watches for changes in the `_worker.js` file, but not for changes in any of its imported dependencies. This means that given the following `_worker.js` file

```
import { graham } from "./graham-the-dog";
export default {
    fetch(request, env) {
	return new Response(graham)
    }
}
```

`pages dev` will reload for any changes in the `_worker.js` file itself, but not for any changes in `graham-the-dog.js`, which is its dependency.

Similarly, `pages dev` will not reload for any changes in non-JS module imports, such as wasm/html/binary module imports.

This commit addresses all the aforementioned issues.

Fixes #4824

## Solutions we considered
Our original solution was implemented using esbuild's `watch` mode (see https://github.com/cloudflare/workers-sdk/pull/6150/commits/dd6cd19707f2070d7dba63e8d79872fce9887d82), which basically meant just enabling the `watch` flag in `buildRawWorker()`. Turned out though (via failed test in CI), that this solution introduced a regression for a frameworks corner case...the deletion of `_worker.js` itself (please see https://github.com/cloudflare/workers-sdk/pull/4877 and https://github.com/cloudflare/workers-sdk/issues/3886). 

When used in the context of frameworks, the assets directory served by `pages dev` gets rebuilt (removed + added back with updated assets) by a separate process, external to wrangler (I'm no expert here on the exact details, but I assume this more or less to be the case). This affects key Pages files, such as `_worker.js` and `_routes.json`, which get deleted/added back as part of that assets directory, which in turn, triggers a `_worker.js` rebuild. If the `_worker.js` re-build is triggered before the external process gets a chance to add the file back, `pages dev` will show a bunch of errors in the terminal, which, while accurate, seem to be confusing/annoying for our users. This was fixed by https://github.com/cloudflare/workers-sdk/pull/4877 by completely turning off Worker rebuilds for chokidar "unlink` events.

Unfortunately, the behaviour for this particular use case is very hard to port to esbuild watch mode, since esbuild gives us very limited information wrt to what changed ....so here we are 😢 

This lead us to the second implementation (see https://github.com/cloudflare/workers-sdk/pull/6150/commits/5c8fcf02018607b0308e595dd63f69ec33274756), which, [similar to Functions](https://github.com/cloudflare/workers-sdk/pull/6022), delegates the entirety of the watch mechanism to `chokidar`. 

## How we tested the changes
This PR comes with e2e tests, but in addition to them, we've manually tested these changes in the following scenarios:

- change/remove/add `_worker.js`
- change in external dependency (eg import { ADD } from "../<dir>/**/<dir>/add")
- change in external module (eg import html from "../<dir>/**/<dir>/my-html.html")
- adding/removing external dependencies
- errors in `_worker.js`
- errors in dependencies

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [x] Required / Maybe required
  - [ ] Not required because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): 
  - [x] Not necessary because: watch mode is not documented

### CODE REVIEW

File: packages/wrangler/src/pages/functions/buildWorker.ts
Comment (andyjessop) [inline packages/wrangler/src/pages/functions/buildWorker.ts:376]: Is this missing a comment?

File: packages/wrangler/src/pages/dev.ts
Comment (andyjessop) [inline packages/wrangler/src/pages/dev.ts:443]: 👍 Fail fast makes a lot of sense here.

File: packages/wrangler/src/pages/functions/buildWorker.ts
Comment (andyjessop) [inline packages/wrangler/src/pages/functions/buildWorker.ts:384]: Nit: could this be `wrangler-dependent-modules-writer`? The word `additional` might be misleading.

**Review by andyjessop — APPROVED** [review]
Looks great! Just some minor comments, but nothing blocking.

File: packages/wrangler/src/pages/functions/buildWorker.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/functions/buildWorker.ts:376]: oh yeah...I was meaning to explain why we needed this plugin to begin with. Thank you for catching that!

File: packages/wrangler/src/pages/functions/buildWorker.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/functions/buildWorker.ts:384]: oh, much better than the name I went for! I like it! I like it a lot!

File: packages/wrangler/src/pages/dev.ts
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/dev.ts:405]: TODO @CarmenPopoviciu add comment

File: packages/wrangler/src/pages/dev.ts
Comment (CarmenPopoviciu) [inline packages/wrangler/src/pages/dev.ts:534]: are we OK with this approach? and if yes, what's a realistic value for `delayMS`. I have abs no feel for it

File: packages/wrangler/src/pages/dev.ts
Comment (penalosa) [inline packages/wrangler/src/pages/dev.ts:414]: Is there a reason to have these as arrays given the functions dir implementation uses `Set`s?

File: packages/wrangler/src/pages/dev.ts
Comment (penalosa) [inline packages/wrangler/src/pages/dev.ts:1]: Can we generalise any of this logic across functions and _worker.js?

File: packages/wrangler/src/pages/dev.ts
Comment (penalosa) [inline packages/wrangler/src/pages/dev.ts:504]: Why would this otherwise build once per typed character? Surely this would only trigger on file save?

File: packages/wrangler/src/pages/dev.ts
Comment (penalosa) [reply packages/wrangler/src/pages/dev.ts:534]: This feels a bit off to me, and I think framework users would probably be surprised that that their rebuilt changes take 5 seconds to show up. What about instead watching the parent directory of `_worker.js`, so we can immediately react to the creation event for `_worker.js`?

File: packages/wrangler/src/pages/dev.ts
Comment (penalosa) [inline packages/wrangler/src/pages/dev.ts:546]: Is this sentence complete?

File: packages/wrangler/src/pages/dev.ts
Comment (penalosa) [inline packages/wrangler/src/pages/dev.ts:549]: Could we lean on esbuild's error formatting here?

File: packages/wrangler/src/pages/dev.ts
Comment (andyjessop) [reply packages/wrangler/src/pages/dev.ts:534]: Why rebuild at all, if they have deleted? Seems like this behaviour would be unexpected.

Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:534]: then perhaps the better question here is do we want to rebuild if `_worker.js` has been deleted? For me it was unexpected that nothing happened when I deleted the file. But maybe that's just me?

Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:534]: > users would probably be surprised that that their rebuilt changes take 5 seconds to show up
to clarify this will only happen when the `_worker.js` file is deleted, **not** on every change

> What about instead watching the parent directory of _worker.js
I might be missing what you have in mind, but how would this solve the problem at hand? The issue is not reacting to the adding of the file, which this implementation already does, it's the adding it "immediately" (not sure what that actually looks like in terms of `ms`) after it was deleted, and either debouncing the `unlink` event handler, or bypassing it completely.

File: packages/wrangler/src/pages/dev.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:414]: this was me trying to simplify what I had in the `Functions` implementation.  That one is a tad more verbose because of the `if...else` clauses, as opposed to this one, which does a lot of array ops chaining, which is why sets were no longer necessary. 

I hesitated writing the code this, because IMO the `Functions` one is a tad more readable...but then again this whole branching of `isUsingWorkerjs` | `isUsingFunctions` | `isUsingWorkerjsDirectory` is getting harder and harder to read :/

File: packages/wrangler/src/pages/dev.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:1]: there was a failed attempt, but reverted because ultimately it felt like the extra layer of indirection adds more complexity than it solves. But let me look at this again and see what I can do

File: packages/wrangler/src/pages/dev.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:504]: indeed, that was not very accurate of me to say. I'll rephrase

File: packages/wrangler/src/pages/dev.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:546]: oeps 😅

File: packages/wrangler/src/pages/dev.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:405]: ✅

File: packages/wrangler/src/pages/dev.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:504]: ✅

File: packages/wrangler/src/pages/dev.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:546]: ✅

File: packages/wrangler/src/pages/dev.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:549]: as I was looking into this, I realised that the `buildNotifierPlugin` already does some logging for us if there were any build errors/warnings. Re-throwing the error here only duplicates that message in a less aesthetic format :/ 👇 
Screenshot 2024-07-03 at 13 44 24

So I'm wondering if it makes sense to output the error again. Would there be any use case where the build err would **not** go through the `buildNotifier` plugin? I'd imagine not. In which case, should we limit this error to a custom message, smth like "failed to build _worker.js", and defer to the notifier plugin to do the error printing?

File: packages/wrangler/src/pages/dev.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts]: refactored Functions to be in line with this PR here: https://github.com/cloudflare/workers-sdk/pull/6150/commits/536836fcd5791a58c3f9b7297b36eb391cdd528c

File: packages/wrangler/src/pages/dev.ts
Comment (penalosa) [reply packages/wrangler/src/pages/dev.ts:535]: Maybe I'd misunderstand, but I thought the problem this was solving was that the `_worker.js` is deleted by some frameworks on every reload?

Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:535]: that is correct yes. and because said frameworks also add the file back, we'd ideally want to either:
- delay building the worker until after the file is added back (which is what this PR does)
- or by-pass rebuilding the worker completely whenever worker file gets deleted (which is how it currently works)

your proposal to react immediately to the `addFile` event (by watching the parent folder), is something that is already accomplished in the crr PR by watching the file itself. The problem in both cases would remain that we'd get all sorts of compile errors when attempting to build the worker after the file was deleted, unless we do one of the 👆

Comment (penalosa) [reply packages/wrangler/src/pages/dev.ts:535]: What I'm concerned about is:
- Framework deletes `_worker.js` file
- We delay building for 5 seconds
- Framework re-adds `_worker.js` file as part of the build process
- User is waiting for 5 seconds for the build to trigger

Would that not happen?

File: packages/wrangler/src/pages/dev.ts
Comment (penalosa) [reply packages/wrangler/src/pages/dev.ts:1]: Makes sense, let's keep things as-is!

**Review by penalosa — APPROVED** [review]

**Comment by penalosa** [comment]
Approved pending https://github.com/cloudflare/workers-sdk/pull/6150#discussion_r1664328291, which I'm likely misunderstanding

File: packages/wrangler/src/pages/dev.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:535]: yes that can happen if the framework adds the file back before those 5s have elapsed. So say file is deleted, then added after 1s...the user needs to wait for 4 more seconds before we trigger a rebuild. But this is why I was saying that I have zero feels abt how long this delay should be. We can set it to whatever we want.

Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:535]: At the same time, if the framework re-adds the file after 7s, we trigger a rebuild after 5s, which will correctly error because the file does not exist, then trigger another build after 2s when the file was re-added, which, assuming all's well, will compile successfully. So in this case, even with a debounce in place, we'd still run into rebuilding a worker file that does not exist, which is what users don't want.

Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:535]: So perhaps, instead of trying to get that delay right, which I'm guessing is impossible, cause the frameworks build process duration is relative to the project size, maybe keep the current behaviour and nor rebuild whenever the file is deleted? I still find the lack of feedback odd, but it might possibly be just me

Comment (penalosa) [reply packages/wrangler/src/pages/dev.ts:535]: > maybe keep the current behaviour and not rebuild whenever the file is deleted?
I think this might be best, even if it is a bit confusing, since frameworks seem to rely on this behaviour. Maybe we could log a small warning to the terminal?

File: packages/wrangler/src/pages/dev.ts
Comment (CarmenPopoviciu) [reply packages/wrangler/src/pages/dev.ts:550]: ✅

**Pull request merged**
